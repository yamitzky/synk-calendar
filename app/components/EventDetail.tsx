import parse from 'html-react-parser'
import { twMerge } from 'tailwind-merge'
import type { CalendarEvent } from '~/domain/calendar'
import useLocale from '~/hooks/useLocale'

type Props = {
  className?: string
} & Pick<CalendarEvent, 'title' | 'start' | 'end' | 'description' | 'location' | 'people' | 'conference'>

export const EventDetail = ({ title, start, end, description, location, people, conference, className }: Props) => {
  const locale = useLocale()
  return (
    <div className={twMerge('space-y-2', className)}>
      <p className="font-bold text-lg">{title}</p>
      <p>{formatRange(locale, start, end)}</p>
      {conference && (
        <p>
          <a href={conference.url} target="_blank" rel="noreferrer" className="">
            {conference.name}
          </a>
        </p>
      )}
      {location && <Location location={location} />}
      {people && (
        <div>
          {people.map((attendee) => (
            <p key={`${attendee.displayName}-${attendee.email}`}>
              {attendee.displayName} &lt;{attendee.email}&gt;
            </p>
          ))}
        </div>
      )}
      {description && <p>{parse(description)}</p>}
    </div>
  )
}

const Location = ({ location }: { location: string }) => {
  let innerContent: React.ReactNode = location
  if (location.startsWith('https://') || location.startsWith('http://')) {
    const domain = new URL(location).hostname
    innerContent = (
      <a href={location} target="_blank" rel="noreferrer">
        {domain}
      </a>
    )
  }
  return <p className="overflow-ellipsis whitespace-nowrap overflow-x-hidden">{innerContent}</p>
}

function formatRange(locale: string | undefined, start: string | null, end: string | null): string {
  const format = new Intl.DateTimeFormat(locale, {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    weekday: 'short',
    hour: 'numeric',
    minute: 'numeric',
  })

  if (start && end) {
    const startDate = new Date(start)
    const endDate = new Date(end)
    return format.formatRange(startDate, endDate)
  }
  if (start) {
    const startDate = new Date(start)
    return format.format(startDate)
  }
  return ''
}
