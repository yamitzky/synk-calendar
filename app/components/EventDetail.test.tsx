import { render, screen } from '@testing-library/react'
import { describe, expect, it } from 'vitest'
import { EventDetail } from './EventDetail'

describe('EventDetail', () => {
  it('renders correctly with all props', () => {
    const { container } = render(
      <EventDetail
        title="Test Event"
        start="2023-05-01T10:00:00"
        end="2023-05-01T11:00:00"
        description="This is a test event"
        location="Test Location"
        people={[{ displayName: 'John Doe', email: 'john@example.com', organizer: false }]}
        conference={{ name: 'Zoom Meeting', url: 'https://zoom.us/j/123456789' }}
      />,
    )

    expect(container).toMatchSnapshot()
    expect(screen.getByText('Test Event')).toBeInTheDocument()
  })

  it('renders correctly with minimal props', () => {
    const { container } = render(
      <EventDetail title="Minimal Event" start="2023-05-01T10:00:00" end="2023-05-01T11:00:00" />,
    )

    expect(container).toMatchSnapshot()
    expect(screen.getByText('Minimal Event')).toBeInTheDocument()
  })

  it('renders location as a link when it is a URL', () => {
    const { container } = render(
      <EventDetail
        title="URL Event"
        start="2023-05-01T10:00:00"
        end="2023-05-01T11:00:00"
        location="https://example.com"
      />,
    )

    const locationLink = screen.getByText('example.com')
    expect(locationLink).toHaveAttribute('href', 'https://example.com')
    expect(container).toMatchSnapshot()
  })
})
