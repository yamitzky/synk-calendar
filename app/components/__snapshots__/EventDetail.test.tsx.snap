// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EventDetail > renders correctly with all props 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <p
      class="font-bold text-lg"
    >
      Test Event
    </p>
    <p>
      2023/5/1(月) 10時00分～11時00分
    </p>
    <p>
      <a
        class=""
        href="https://zoom.us/j/123456789"
        rel="noreferrer"
        target="_blank"
      >
        Zoom Meeting
      </a>
    </p>
    <p
      class="overflow-ellipsis whitespace-nowrap overflow-x-hidden"
    >
      Test Location
    </p>
    <div>
      <p>
        John Doe
         &lt;
        john@example.com
        &gt;
      </p>
    </div>
    <p>
      This is a test event
    </p>
  </div>
</div>
`;

exports[`EventDetail > renders correctly with minimal props 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <p
      class="font-bold text-lg"
    >
      Minimal Event
    </p>
    <p>
      2023/5/1(月) 10時00分～11時00分
    </p>
  </div>
</div>
`;

exports[`EventDetail > renders location as a link when it is a URL 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <p
      class="font-bold text-lg"
    >
      URL Event
    </p>
    <p>
      2023/5/1(月) 10時00分～11時00分
    </p>
    <p
      class="overflow-ellipsis whitespace-nowrap overflow-x-hidden"
    >
      <a
        href="https://example.com"
        rel="noreferrer"
        target="_blank"
      >
        example.com
      </a>
    </p>
  </div>
</div>
`;
